src/metabase/sync_database/cached_values.clj

```diff
@@ -0,0 +1,136 @@
+(ns metabase.sync-database.cached-values
+  "Functions discover and save the fields in a table."
+  (:require [clojure.tools.logging :as log]
+            [metabase
+             [driver :as driver]
+             [util :as u]]
+            [metabase.models
+             [field :as field]
+             [field-values :as field-values]
+             [table :as table]]
+            [metabase.sync-database.interface :as i]
+            [schema.core :as schema]
+            [toucan.db :as db]
+            [metabase.db.metadata-queries :as queries]
+            [metabase.sync-database.classify :as classify]
+            [metabase.sync-database.infer-special-type :as infer-special-type]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;; Drivers use these to get field values
+
+(def ^:private ^:const ^Integer field-values-entry-max-length
+  "The maximum character length for a stored `FieldValues` entry."
+  100)
+
+(def ^:private ^:const ^Integer field-values-total-max-length
+  "Maximum total length for a FieldValues entry (combined length of all values for the field)."
+  (* classify/low-cardinality-threshold field-values-entry-max-length))
+
+(defn field-values-below-low-cardinality-threshold? [non-nil-values]
+  (and (<= (count non-nil-values) classify/low-cardinality-threshold)
+      ;; very simple check to see if total length of field-values exceeds (total values * max per value)
+       (let [total-length (reduce + (map (comp count str) non-nil-values))]
+         (<= total-length field-values-total-max-length))))
+
+(defn test-for-cardinality?
+  "Should FIELD should be tested for cardinality?"
+  [{:keys [base_type] :as field}]
+  (or (field-values/field-should-have-field-values? field)
+      (and (not (isa? base_type :type/DateTime))
+           (not (isa? base_type :type/Collection))
```

Comment:
from a readability standpoint it really might just make sense to derive the types in question here from a one-off namespaced type and then use that for the checks. e.g. something like

```clojure
(derive ::skip-cardinality-testing :type/DateTime)
(derive ::skip-cardinality-testing :type/Collection)
...
;; Don't test base types that derive from ::skip-cardinality-testing
(not (isa? base_type ::skip-cardinality-testing))
  ...)
```